File: apply_patch.py
"""MyBeatSaberStats パッチ適用プログラム。

配布構成:
  MyBeatSaberStats/
  ├── MyBeatSaberStatsPlayer.exe
  ├── apply_patch.exe        ← このプログラム
  ├── _internal/             ← メインアプリの _internal（更新対象）
  │   ├── lib/mybeatsaberstats/
  │   ├── resources/
  │   └── version.json
  └── patch/                 ← パッチ内容（同フォルダに展開しておく）
      ├── lib/
      │   └── mybeatsaberstats/
      │       ├── *.py
      │       └── collector/*.py
      ├── resources/          ← アイコン等（存在する場合のみ適用）
      │   └── *.ico, *.png, ...
      └── version.json

使い方:
  1. patch/ フォルダを apply_patch.exe と同じフォルダに置く
  2. apply_patch.exe を実行してパッチを適用する
  3. メインアプリを再起動する
"""

from __future__ import annotations

import json
import shutil
import sys
from pathlib import Path

from PySide6.QtCore import Qt, QThread, Signal
from PySide6.QtWidgets import (
    QApplication, QDialog, QHBoxLayout, QLabel,
    QMessageBox, QProgressBar, QPushButton, QVBoxLayout,
)


# ------------------------------------------------------------------ #
#  パスユーティリティ
# ------------------------------------------------------------------ #

def _patcher_dir() -> Path:
    """apply_patch.exe があるフォルダ（frozen 時は exe の親、開発時はスクリプトの親）。"""
    if getattr(sys, "frozen", False):
        return Path(sys.executable).parent
    return Path(__file__).resolve().parent


def _patch_dir() -> Path:
    return _patcher_dir() / "patch"


def _internal_dir() -> Path:
    return _patcher_dir() / "_internal"


def _read_version(version_json: Path) -> str | None:
    try:
        data = json.loads(version_json.read_text("utf-8"))
        ver = data.get("version")
        return str(ver).lstrip("v") if ver else None
    except Exception:
        return None


# ------------------------------------------------------------------ #
#  バックグラウンドスレッド
# ------------------------------------------------------------------ #

class _ApplyThread(QThread):
    progress = Signal(str)
    finished = Signal()
    error    = Signal(str)

    def run(self) -> None:
        try:
            patch_dir    = _patch_dir()
            internal_dir = _internal_dir()
            lib_src      = patch_dir    / "lib"
            lib_dest     = internal_dir / "lib"
            res_src      = patch_dir    / "resources"
            res_dest     = internal_dir / "resources"
            ver_src      = patch_dir    / "version.json"
            ver_dest     = internal_dir / "version.json"

            # 1. 既存 lib を削除
            self.progress.emit("既存の lib フォルダを削除中...")
            if lib_dest.exists():
                shutil.rmtree(lib_dest)

            # 2. patch/lib を _internal/lib にコピー
            self.progress.emit("新しい lib フォルダをコピー中...")
            shutil.copytree(lib_src, lib_dest)

            # 3. patch/resources を _internal/resources にコピー (存在する場合のみ)
            if res_src.exists():
                self.progress.emit("resources フォルダを更新中...")
                if res_dest.exists():
                    shutil.rmtree(res_dest)
                shutil.copytree(res_src, res_dest)

            # 4. version.json を上書き
            self.progress.emit("version.json を更新中...")
            shutil.copy2(ver_src, ver_dest)

            self.finished.emit()
        except Exception as e:
            self.error.emit(str(e))


# ------------------------------------------------------------------ #
#  ダイアログ
# ------------------------------------------------------------------ #

class PatchDialog(QDialog):
    def __init__(self) -> None:
        super().__init__()
        self.setWindowTitle("MyBeatSaberStats パッチ適用")
        self.setMinimumWidth(440)
        self.setWindowFlags(
            self.windowFlags() & ~Qt.WindowType.WindowContextHelpButtonHint
        )
        self._thread: _ApplyThread | None = None

        layout = QVBoxLayout(self)

        patch_dir    = _patch_dir()
        internal_dir = _internal_dir()

        # バリデーション
        ok, err_msg = self._validate(patch_dir, internal_dir)
        if not ok:
            layout.addWidget(QLabel(f"<b>エラー:</b><br>{err_msg}"))
            close_btn = QPushButton("閉じる")
            close_btn.clicked.connect(self.reject)
            layout.addWidget(close_btn)
            return

        # バージョン表示
        current_ver = _read_version(internal_dir / "version.json") or "不明"
        new_ver     = _read_version(patch_dir    / "version.json") or "不明"
        layout.addWidget(QLabel(
            f"<b>パッチを適用しますか？</b><br><br>"
            f"現在のバージョン: <b>v{current_ver}</b><br>"
            f"適用後のバージョン: <b>v{new_ver}</b><br><br>"
            f"<small>※ 適用前にアプリを終了してください。</small>"
        ))

        # 進捗表示
        self._progress_label = QLabel("")
        self._progress_bar   = QProgressBar()
        self._progress_bar.setRange(0, 0)
        self._progress_label.hide()
        self._progress_bar.hide()
        layout.addWidget(self._progress_label)
        layout.addWidget(self._progress_bar)

        # ボタン行
        btn_row = QHBoxLayout()
        self._apply_btn  = QPushButton("適用する")
        self._cancel_btn = QPushButton("キャンセル")
        self._apply_btn.setDefault(True)
        self._apply_btn.clicked.connect(self._start_apply)
        self._cancel_btn.clicked.connect(self.reject)
        btn_row.addStretch(1)
        btn_row.addWidget(self._cancel_btn)
        btn_row.addWidget(self._apply_btn)
        layout.addLayout(btn_row)

    # ------------------------------------------------------------------

    @staticmethod
    def _validate(patch_dir: Path, internal_dir: Path) -> tuple[bool, str]:
        if not patch_dir.exists():
            return False, (
                f"patch フォルダが見つかりません。<br>"
                f"apply_patch.exe と同じフォルダに patch/ フォルダを置いてください。<br>"
                f"<small>{patch_dir}</small>"
            )
        if not (patch_dir / "version.json").exists():
            return False, "patch/version.json が見つかりません。"
        if not (patch_dir / "lib").exists():
            return False, "patch/lib フォルダが見つかりません。"
        if not internal_dir.exists():
            return False, (
                f"_internal フォルダが見つかりません。<br>"
                f"メインアプリと同じフォルダで実行してください。<br>"
                f"<small>{internal_dir}</small>"
            )
        if not (internal_dir / "version.json").exists():
            return False, "_internal/version.json が見つかりません。"
        return True, ""

    def _start_apply(self) -> None:
        self._apply_btn.setEnabled(False)
        self._cancel_btn.setEnabled(False)
        self._progress_label.show()
        self._progress_bar.show()

        self._thread = _ApplyThread()
        self._thread.progress.connect(self._on_progress)
        self._thread.finished.connect(self._on_finished)
        self._thread.error.connect(self._on_error)
        self._thread.start()

    def _on_progress(self, msg: str) -> None:
        self._progress_label.setText(msg)
        QApplication.processEvents()

    def _on_finished(self) -> None:
        self._progress_bar.setRange(0, 1)
        self._progress_bar.setValue(1)
        self._progress_label.setText("完了しました！")
        new_ver = _read_version(_patch_dir() / "version.json") or "不明"
        QMessageBox.information(
            self,
            "パッチ適用完了",
            f"v{new_ver} へのパッチ適用が完了しました。\n"
            "アプリを起動すると新しいバージョンが有効になります。",
        )
        self.accept()

    def _on_error(self, msg: str) -> None:
        self._progress_bar.setRange(0, 1)
        self._progress_bar.setValue(0)
        self._progress_label.setText("エラーが発生しました。")
        self._apply_btn.setEnabled(True)
        self._cancel_btn.setEnabled(True)
        QMessageBox.critical(self, "エラー", f"パッチ適用中にエラーが発生しました:\n{msg}")


# ------------------------------------------------------------------ #
#  エントリポイント
# ------------------------------------------------------------------ #

def main() -> None:
    app = QApplication(sys.argv)
    dlg = PatchDialog()
    dlg.show()
    sys.exit(app.exec())


if __name__ == "__main__":
    main()

File: apply_patch.cs
// MyBeatSaberStats パッチ適用プログラム (C# WinForms)
//
// 配布構成:
//   MyBeatSaberStats/
//   ├── MyBeatSaberStatsPlayer.exe
//   ├── apply_patch.exe        ← このプログラム（patch.zip を埋め込み可）
//   ├── _internal/             ← メインアプリの _internal（更新対象）
//   │   ├── lib/mybeatsaberstats/
//   │   └── version.json
//   └── patch/                 ← パッチ内容（同フォルダに展開しておく、任意）
//       ├── lib/
//       │   └── mybeatsaberstats/
//       │       ├── *.py
//       │       └── collector/*.py
//       ├── PySide6/
//       │   ├── QtSvg.pyd
//       │   └── Qt6Svg.dll
//       ├── resources/
//       │   └── *
//       └── version.json
//
// ビルド: .NET Framework 4.x の csc.exe で単一 EXE にコンパイル
//   csc.exe /nologo /target:winexe /out:apply_patch.exe
//           /reference:System.Windows.Forms.dll /reference:System.Drawing.dll
//           apply_patch.cs

using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.IO;
using System.IO.Compression;
using System.Reflection;
using System.Text.RegularExpressions;
using System.Windows.Forms;

static class Program
{
    [STAThread]
    static void Main()
    {
        Application.EnableVisualStyles();
        Application.SetCompatibleTextRenderingDefault(false);
        Application.Run(new PatchDialog());
    }
}

sealed class PatchDialog : Form
{
    private static readonly string[] DefaultManagedRoots = { "lib", "PySide6", "resources" };

    // UI パーツ
    private readonly Label        _infoLabel;
    private readonly Label        _progressLabel;
    private readonly ProgressBar  _progressBar;
    private readonly Button       _applyBtn;
    private readonly Button       _cancelBtn;

    // パス
    private readonly string _patchDir;
    private readonly string _internalDir;
    private readonly string _tempPatchDir;

    public PatchDialog()
    {
        string exe       = System.Reflection.Assembly.GetExecutingAssembly().Location;
        string patcherDir = Path.GetDirectoryName(exe);
        _internalDir = Path.Combine(patcherDir, "_internal");
        _patchDir    = ResolvePatchDir(patcherDir, out _tempPatchDir);

        // ---------- フォーム設定 ----------
        this.Text            = "MyBeatSaberStats パッチ適用";
        this.FormBorderStyle = FormBorderStyle.FixedDialog;
        this.MaximizeBox     = false;
        this.MinimizeBox     = false;
        this.StartPosition   = FormStartPosition.CenterScreen;
        this.ClientSize      = new System.Drawing.Size(460, 220);
        this.Padding         = new Padding(16);

        // ---------- コントロール ----------
        _infoLabel = new Label
        {
            AutoSize = false,
            Size     = new System.Drawing.Size(428, 90),
            Location = new System.Drawing.Point(16, 16),
        };

        _progressLabel = new Label
        {
            AutoSize = false,
            Size     = new System.Drawing.Size(428, 20),
            Location = new System.Drawing.Point(16, 112),
            Visible  = false,
        };

        _progressBar = new ProgressBar
        {
            Size     = new System.Drawing.Size(428, 18),
            Location = new System.Drawing.Point(16, 138),
            Style    = ProgressBarStyle.Marquee,
            Visible  = false,
        };

        _cancelBtn = new Button
        {
            Text     = "キャンセル",
            Size     = new System.Drawing.Size(88, 30),
            Location = new System.Drawing.Point(252, 172),
        };

        _applyBtn = new Button
        {
            Text     = "適用する",
            Size     = new System.Drawing.Size(88, 30),
            Location = new System.Drawing.Point(356, 172),
        };

        _applyBtn.Click  += OnApply;
        _cancelBtn.Click += (s, e) => Application.Exit();

        this.AcceptButton = _applyBtn;
        this.CancelButton = _cancelBtn;

        this.Controls.AddRange(new Control[]
        {
            _infoLabel, _progressLabel, _progressBar, _cancelBtn, _applyBtn
        });

        // ---------- 内容セットアップ ----------
        string errMsg = ValidatePatch();
        if (errMsg != null)
        {
            _infoLabel.Text      = "エラー:\r\n" + errMsg;
            _applyBtn.Enabled    = false;
            return;
        }

        string currentVer = ReadVersion(Path.Combine(_internalDir, "version.json")) ?? "不明";
        string newVer     = ReadVersion(Path.Combine(_patchDir,    "version.json")) ?? "不明";
        _infoLabel.Text =
            "パッチを適用しますか？\r\n\r\n" +
            "現在のバージョン : v" + currentVer + "\r\n" +
            "適用後のバージョン: v" + newVer     + "\r\n\r\n" +
            "※ 適用前にメインアプリを終了してください。";
    }

    protected override void OnClosed(EventArgs e)
    {
        base.OnClosed(e);
        CleanupTemporaryPatchDir();
    }

    // ------------------------------------------------------------------
    //  バリデーション
    // ------------------------------------------------------------------

    private string ValidatePatch()
    {
        string[] managedRoots = LoadManagedInternalRoots();
        if (!Directory.Exists(_patchDir))
            return "パッチ内容が見つかりません。\r\n" +
                   "apply_patch.exe と同じフォルダに patch/ フォルダを置くか、\r\n" +
                   "埋め込みパッチ付きの apply_patch.exe を使用してください。\r\n" +
                   "(" + _patchDir + ")";
        if (!File.Exists(Path.Combine(_patchDir, "version.json")))
            return "patch/version.json が見つかりません。";
        foreach (string root in managedRoots)
        {
            if (!Directory.Exists(Path.Combine(_patchDir, root)))
                return "patch/" + root + " フォルダが見つかりません。";
        }
        if (!Directory.Exists(_internalDir))
            return "_internal フォルダが見つかりません。\r\n" +
                   "メインアプリと同じフォルダで実行してください。\r\n" +
                   "(" + _internalDir + ")";
        if (!File.Exists(Path.Combine(_internalDir, "version.json")))
            return "_internal/version.json が見つかりません。";
        return null;
    }

    // ------------------------------------------------------------------
    //  バージョン読み込み（System.Text.Json 不要・正規表現で抽出）
    // ------------------------------------------------------------------

    private static string ReadVersion(string path)
    {
        try
        {
            string text = File.ReadAllText(path, System.Text.Encoding.UTF8);
            var m = Regex.Match(text, @"""version""\s*:\s*""([^""]+)""");
            return m.Success ? m.Groups[1].Value.TrimStart('v') : null;
        }
        catch
        {
            return null;
        }
    }

    // ------------------------------------------------------------------
    //  パッチ適用（BackgroundWorker で非同期）
    // ------------------------------------------------------------------

    private void OnApply(object sender, EventArgs e)
    {
        _applyBtn.Enabled  = false;
        _cancelBtn.Enabled = false;
        _progressLabel.Visible = true;
        _progressBar.Visible   = true;

        var worker = new BackgroundWorker();
        worker.DoWork             += DoApply;
        worker.RunWorkerCompleted += OnApplyCompleted;
        worker.RunWorkerAsync();
    }

    private void DoApply(object sender, DoWorkEventArgs e)
    {
        try
        {
            string[] managedRoots = LoadManagedInternalRoots();
            string verSrc  = Path.Combine(_patchDir,    "version.json");
            string verDest = Path.Combine(_internalDir, "version.json");

            foreach (string root in managedRoots)
            {
                string src = Path.Combine(_patchDir, root);
                string dest = Path.Combine(_internalDir, root);

                SetProgress("既存の " + root + " フォルダを削除中...");
                if (Directory.Exists(dest))
                    Directory.Delete(dest, recursive: true);

                SetProgress("新しい " + root + " フォルダをコピー中...");
                CopyDirectory(src, dest);
            }

            SetProgress("version.json を更新中...");
            File.Copy(verSrc, verDest, overwrite: true);
        }
        catch (Exception ex)
        {
            e.Result = ex.Message;
        }
    }

    private void OnApplyCompleted(object sender, RunWorkerCompletedEventArgs e)
    {
        _progressBar.Visible = false;

        string errorMsg = e.Result as string;
        if (errorMsg != null)
        {
            _progressLabel.Text = "エラー: " + errorMsg;
            MessageBox.Show(
                "エラーが発生しました:\r\n" + errorMsg,
                "エラー",
                MessageBoxButtons.OK,
                MessageBoxIcon.Error);
            _cancelBtn.Enabled = true;
        }
        else
        {
            _progressLabel.Text = "完了！アプリを再起動してください。";
            MessageBox.Show(
                "パッチが正常に適用されました。\r\nアプリを再起動してください。",
                "完了",
                MessageBoxButtons.OK,
                MessageBoxIcon.Information);
            Application.Exit();
        }
    }

    // ------------------------------------------------------------------
    //  ユーティリティ
    // ------------------------------------------------------------------

    private void SetProgress(string msg)
    {
        if (InvokeRequired)
            Invoke((Action)(() => _progressLabel.Text = msg));
        else
            _progressLabel.Text = msg;
    }

    private static string ResolvePatchDir(string patcherDir, out string tempPatchDir)
    {
        string patchDir = Path.Combine(patcherDir, "patch");
        tempPatchDir = null;
        if (Directory.Exists(patchDir))
            return patchDir;

        string resourceName = FindEmbeddedPatchResourceName();
        if (resourceName == null)
            return patchDir;

        tempPatchDir = Path.Combine(Path.GetTempPath(), "MyBeatSaberStatsPatch_" + Guid.NewGuid().ToString("N"));
        Directory.CreateDirectory(tempPatchDir);
        string zipPath = Path.Combine(tempPatchDir, "patch_payload.zip");

        using (Stream stream = Assembly.GetExecutingAssembly().GetManifestResourceStream(resourceName))
        {
            if (stream == null)
                return patchDir;
            using (FileStream fs = File.Create(zipPath))
                stream.CopyTo(fs);
        }

        string extractDir = Path.Combine(tempPatchDir, "patch");
        ZipFile.ExtractToDirectory(zipPath, extractDir);
        return extractDir;
    }

    private static string FindEmbeddedPatchResourceName()
    {
        foreach (string name in Assembly.GetExecutingAssembly().GetManifestResourceNames())
        {
            if (name.EndsWith("patch_payload.zip", StringComparison.OrdinalIgnoreCase))
                return name;
        }
        return null;
    }

    private void CleanupTemporaryPatchDir()
    {
        try
        {
            if (!string.IsNullOrEmpty(_tempPatchDir) && Directory.Exists(_tempPatchDir))
                Directory.Delete(_tempPatchDir, recursive: true);
        }
        catch
        {
        }
    }

    private string[] LoadManagedInternalRoots()
    {
        string configPath = Path.Combine(_patchDir, "resources", "update_targets.json");
        try
        {
            if (!File.Exists(configPath))
                return DefaultManagedRoots;

            string text = File.ReadAllText(configPath, System.Text.Encoding.UTF8);
            Match arrayMatch = Regex.Match(
                text,
                @"""internal_sync_dirs""\s*:\s*\[(.*?)\]",
                RegexOptions.Singleline);
            if (!arrayMatch.Success)
                return DefaultManagedRoots;

            var roots = new List<string>();
            MatchCollection matches = Regex.Matches(arrayMatch.Groups[1].Value, @"""([^""]+)""");
            foreach (Match match in matches)
            {
                string value = match.Groups[1].Value.Trim();
                if (value.Length > 0)
                    roots.Add(value);
            }
            return roots.Count > 0 ? roots.ToArray() : DefaultManagedRoots;
        }
        catch
        {
            return DefaultManagedRoots;
        }
    }

    private static void CopyDirectory(string src, string dest)
    {
        Directory.CreateDirectory(dest);
        foreach (string file in Directory.GetFiles(src))
            File.Copy(file, Path.Combine(dest, Path.GetFileName(file)), overwrite: true);
        foreach (string dir in Directory.GetDirectories(src))
            CopyDirectory(dir, Path.Combine(dest, Path.GetFileName(dir)));
    }
}
